// @vladislav_vasilyev/m0603/m0603.js
const fs = require('fs');
const path = require('path');
const nodemailer = require('nodemailer');

let mailConfig;
try {
    mailConfig = require('../mailconfig.json');
} catch (e) {
    console.error('Не удалось загрузить mailconfig.json:', e && e.message ? e.message : e);
    process.exit(1);
}

if (!mailConfig || !mailConfig.host || !mailConfig.port || !mailConfig.auth || !mailConfig.auth.user) {
    console.error('mailconfig.json должен содержать host, port и auth.user');
    process.exit(1);
}

// Пароль берем либо из env `SMTP_PASS`, либо из файла `nodemailer pass` (если он есть).
// Так можно публиковать пакет в npm, не раскрывая пароль.
function resolvePassword() {
    if (process.env.SMTP_PASS) return String(process.env.SMTP_PASS).trim();
    // Пытаемся найти файл в текущей директории приложения (так удобнее после установки из npm)
    const passFileFromCwd = path.join(process.cwd(), 'nodemailer pass');
    if (fs.existsSync(passFileFromCwd)) return fs.readFileSync(passFileFromCwd, 'utf8').trim();

    // На случай учебной сборки: также проверяем рядом с самим модулем
    const passFileFromModuleDir = path.join(__dirname, 'nodemailer pass');
    if (fs.existsSync(passFileFromModuleDir)) return fs.readFileSync(passFileFromModuleDir, 'utf8').trim();
    return null;
}

const pass = resolvePassword();
if (!pass) {
    throw new Error(
        'Не найден SMTP_PASS (env) и/или файл "nodemailer pass". ' +
        'Укажи пароль SMTP в переменной окружения SMTP_PASS.'
    );
}

mailConfig = {
    ...mailConfig,
    auth: {
        ...mailConfig.auth,
        pass,
    },
};

// Фиксированный адрес получателя (как в задании).
const FIXED_EMAIL = mailConfig.auth.user;

const transporter = nodemailer.createTransport(mailConfig);

/**
 * send(text)
 * @param {string} text
 */
async function send(text) {
    const content = String(text);

    return transporter.sendMail({
        from: FIXED_EMAIL,
        to: FIXED_EMAIL,
        subject: '06-03',
        text: content,
        html: `<p>${content.replace(/\n/g, '<br>')}</p>`,
    });
}

module.exports = { send };
